### PR TITLE
datastructures patch

### PR DESCRIPTION
**Pull request by jd-lara**
The breaking change in DataStructures required syntax updates and removal of compatibility with 0.17

### CODE REVIEW

**Comment by jd-lara** [comment]
I will wait until the tests pass and then merge and tag.

**Comment by devmotion** [comment]
Yes, I noticed this (and have a branch with the changes for OrdinaryDiffEq already on my computer). The problem is that there are many cyclic dependencies (e.g., DiffEqCallbacks needs an update as well but it uses OrdinaryDiffEq in its dependencies as well...). That's also a problem with this PR and the reason for the failing tests.

File: src/common_interface/integrator_types.jl [hunk 1]
```diff
@@ -155,8 +155,8 @@ DiffEqBase.postamble!(integrator::AbstractSundialsIntegrator) = nothing
 
 @inline function DiffEqBase.step!(integrator::AbstractSundialsIntegrator)
   if integrator.opts.advance_to_tstop
-    while integrator.tdir*(integrator.t-DataStructures.top(integrator.opts.tstops)) < -1e6eps()
-        tstop = DataStructures.top(integrator.opts.tstops)
+    while integrator.tdir*(integrator.t-DataStructures.first(integrator.opts.tstops)) < -1e6eps()
```
Comment (devmotion) [inline src/common_interface/integrator_types.jl:158]:
```suggestion
    while integrator.tdir * (integrator.t - first(integrator.opts.tstops)) < -1e6eps()
```

File: src/common_interface/integrator_types.jl [hunk 2]
```diff
@@ -155,8 +155,8 @@ DiffEqBase.postamble!(integrator::AbstractSundialsIntegrator) = nothing
...
+        tstop = DataStructures.first(integrator.opts.tstops)
```
Comment (devmotion) [inline src/common_interface/integrator_types.jl:159]:
```suggestion
        tstop = first(integrator.opts.tstops)
```

File: src/common_interface/integrator_types.jl [hunk 3]
```diff
@@ -174,7 +174,7 @@ DiffEqBase.postamble!(integrator::AbstractSundialsIntegrator) = nothing
           integrator.uprev .= integrator.u
       end
       if !isempty(integrator.opts.tstops)
-          tstop = DataStructures.top(integrator.opts.tstops)
+          tstop = DataStructures.first(integrator.opts.tstops)
```
Comment (devmotion) [inline src/common_interface/integrator_types.jl:177]:
```suggestion
          tstop = first(integrator.opts.tstops)
```

File: src/common_interface/integrator_utils.jl [hunk 4]
```diff
@@ -39,7 +39,7 @@ function DiffEqBase.savevalues!(integrator::AbstractSundialsIntegrator,force_sav
   !integrator.opts.save_on && return saved, savedexactly
   uType = eltype(integrator.sol.u)
   while !isempty(integrator.opts.saveat) &&
-    integrator.tdir*DataStructures.top(integrator.opts.saveat) < integrator.tdir*integrator.t
+    integrator.tdir*DataStructures.first(integrator.opts.saveat) < integrator.tdir*integrator.t
```
Comment (devmotion) [inline src/common_interface/integrator_utils.jl:42]:
```suggestion
    integrator.tdir * first(integrator.opts.saveat) < integrator.tdir * integrator.t
```

File: src/common_interface/solve.jl [hunk 5]
```diff
@@ -1153,8 +1153,8 @@ function DiffEqBase.solve!(integrator::AbstractSundialsIntegrator)
     while !isempty(integrator.opts.tstops)
         # Sundials can have floating point issues approaching a tstop if
         # there is a modifying event each
-        while integrator.tdir*(integrator.t-DataStructures.top(integrator.opts.tstops)) < -1e6eps()
-            tstop = DataStructures.top(integrator.opts.tstops)
+        while integrator.tdir*(integrator.t-DataStructures.first(integrator.opts.tstops)) < -1e6eps()
+            tstop = DataStructures.first(integrator.opts.tstops)
```
Comment (devmotion) [inline src/common_interface/solve.jl:1157]:
```suggestion
        while integrator.tdir * (integrator.t - first(integrator.opts.tstops)) < -1e6eps()
            tstop = first(integrator.opts.tstops)
```

File: src/common_interface/solve.jl [hunk 6]
```diff
@@ -1217,7 +1217,7 @@ end
 function handle_tstop!(integrator::AbstractSundialsIntegrator)
     tstops = integrator.opts.tstops
     if !isempty(tstops)
-      if integrator.tdir*(integrator.t-DataStructures.top(integrator.opts.tstops)) > -1e6eps()
+      if integrator.tdir*(integrator.t-DataStructures.first(integrator.opts.tstops)) > -1e6eps()
```
Comment (devmotion) [inline src/common_interface/solve.jl:1220]:
```suggestion
      if integrator.tdir * (integrator.t - first(integrator.opts.tstops)) > -1e6eps()
```

**Comment by ChrisRackauckas** [comment]
The dependencies are only cyclic in the tests, but yeah it's a bit annoying. That's why in other libraries I've split out the tests of downstream features into an isolated environment. We should probably do more of that

**Comment by jd-lara** [comment]
@devmotion I am planning to make a PR with no logic changes to clean up the formatting using Julia Formatting

**Comment by devmotion** [comment]
The suggestions are not related to formatting, my suggestion is just to use `first` instead of `DataStructures.first`.

**Comment by jd-lara** [comment]
@devmotion Sundials.jl uses `import DataStructures` so it needs the predicate. As a practical aspect is easier to find broken dependencies and make it clear to future maintainers which functions come from dependencies which ones are part of the package.

**Comment by devmotion** [comment]
Yes, I saw that you `import DataStructures` and personally I like this approach. But DataStructures just overloads `Base.first` which is always available as `first`, even if you don't import anything.

**Comment by jd-lara** [comment]
@devmotion I see your point. However, in my experience packages have no requirement to maintain these overloading functions and are prone to change exactly as it happened here: `top` used to be a Julia `Base` function that changed to `first` and until now DataStructures made the update in the syntax.  

I have had plenty of headaches with packages in the pre 1.0 stage making this sort of change (or not making it on time) and costing me time later chasing the correct compatibility. 

That said if the preference of the community is to remove the predicate I will respect it. @ChrisRackauckas what do you think?

**Comment by ChrisRackauckas** [comment]
We should use `import` more, but `first` doesn't come from DataStructures.

**Comment by devmotion** [comment]
IMO these problems only occur if packages do not follow SemVer. If you specify compatibility bounds, breaking changes in dependencies shouldn't suddenly break your package.

**Comment by jd-lara** [comment]
> We should use `import` more, but `first` doesn't come from DataStructures.

@devmotion I made the change but added a comment for future reference

**Comment by devmotion** [comment]
As said above, IMO DataStructures is only allowed to remove the `first` overload in a breaking release, hence I don't think you should not be worried about any details of the DataStructures implementation.

**Comment by jd-lara** [comment]
It looks like the issues with the tests are upstream. I will merge to continue with other improvements before tagging.

**Pull request merged**